File: app/src/main/java/me/imzack/app/ender/view/dialog/TypeMarkPatternPickerDialogFragment.java
package me.imzack.app.ender.view.dialog;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.GridView;

import me.imzack.app.ender.R;
import me.imzack.app.ender.model.DataManager;
import me.imzack.app.ender.model.bean.TypeMarkPattern;
import me.imzack.app.ender.util.ResourceUtil;
import me.imzack.app.ender.view.adapter.TypeMarkPatternGridAdapter;

import java.io.Serializable;
import java.util.List;

import butterknife.BindView;

public class TypeMarkPatternPickerDialogFragment extends BaseDialogFragment {

    @BindView(R.id.grid_type_mark_pattern)
    GridView mTypeMarkPatternGrid;

    private static final String ARG_DEFAULT_PATTERN = "default_pattern";
    private static final String ARG_TYPE_MARK_PATTERN_PICKED_LSNR = "type_mark_pattern_picked_lsnr";

    private TypeMarkPattern mTypeMarkPattern;
    private int mPosition = -1;
    private List<TypeMarkPattern> mTypeMarkPatternList;
    private OnTypeMarkPatternPickedListener mOnTypeMarkPatternPickedListener;

    public TypeMarkPatternPickerDialogFragment() {

    }

    public static TypeMarkPatternPickerDialogFragment newInstance(String defaultPattern, OnTypeMarkPatternPickedListener listener) {
        TypeMarkPatternPickerDialogFragment fragment = new TypeMarkPatternPickerDialogFragment();
        Bundle args = new Bundle();
        args.putString(ARG_TITLE_STR, ResourceUtil.getString(R.string.title_dialog_fragment_type_mark_pattern_picker));
        args.putString(ARG_NEU_BTN_STR, ResourceUtil.getString(R.string.button_remove));
        args.putString(ARG_NEG_BTN_STR, ResourceUtil.getString(R.string.button_cancel));
        args.putString(ARG_POS_BTN_STR, ResourceUtil.getString(R.string.button_select));
        args.putString(ARG_DEFAULT_PATTERN, defaultPattern);
        args.putSerializable(ARG_TYPE_MARK_PATTERN_PICKED_LSNR, listener);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        String defaultPattern = null;
        Bundle args = getArguments();
        if (args != null) {
            defaultPattern = args.getString(ARG_DEFAULT_PATTERN);
            mOnTypeMarkPatternPickedListener = (OnTypeMarkPatternPickedListener) args.getSerializable(ARG_TYPE_MARK_PATTERN_PICKED_LSNR);
        }

        mTypeMarkPatternList = DataManager.getInstance().getTypeMarkPatternList();

        mTypeMarkPattern = new TypeMarkPattern();

        if (defaultPattern != null) {
            for (int i = 0; i < mTypeMarkPatternList.size(); i++) {
                TypeMarkPattern typeMarkPattern = mTypeMarkPatternList.get(i);
                if (typeMarkPattern.getPatternFn().equals(defaultPattern)) {
                    mPosition = i;
                    mTypeMarkPattern.setPattern(typeMarkPattern.getPatternFn(), typeMarkPattern.getPatternName());
                    break;
                }
            }
        }

        setNeutralButtonClickListener(new OnButtonClickListener() {
            @Override
            public boolean onClick() {
                if (mOnTypeMarkPatternPickedListener != null) {
                    mOnTypeMarkPatternPickedListener.onTypeMarkPatternPicked(null);
                }
                return true;
            }
        });
        setPositiveButtonClickListener(new OnButtonClickListener() {
            @Override
            public boolean onClick() {
                if (mOnTypeMarkPatternPickedListener != null) {
                    mOnTypeMarkPatternPickedListener.onTypeMarkPatternPicked(mTypeMarkPattern.getPatternFn() == null ? null : mTypeMarkPattern);
                }
                return true;
            }
        });
    }

    @Override
    public View onCreateContentView(LayoutInflater inflater, ViewGroup root) {
        return inflater.inflate(R.layout.dialog_fragment_type_mark_pattern_picker, root, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mTypeMarkPatternGrid.setAdapter(new TypeMarkPatternGridAdapter(mTypeMarkPatternList));

        if (mPosition != -1) {
            mTypeMarkPatternGrid.setItemChecked(mPosition, true);
        }

        mTypeMarkPatternGrid.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                mPosition = position;
                TypeMarkPattern typeMarkPattern = mTypeMarkPatternList.get(mPosition);
                mTypeMarkPattern.setPattern(typeMarkPattern.getPatternFn(), typeMarkPattern.getPatternName());
            }
        });
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mOnTypeMarkPatternPickedListener = null;
    }

    public interface OnTypeMarkPatternPickedListener extends Serializable {
        void onTypeMarkPatternPicked(TypeMarkPattern typeMarkPattern);
    }
}

File: app/src/main/kotlin/me/imzack/app/end/view/dialog/TypeMarkPatternPickerDialogFragment.kt
package me.imzack.app.end.view.dialog

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import kotlinx.android.synthetic.main.dialog_fragment_type_mark_pattern_picker.*
import me.imzack.app.end.R
import me.imzack.app.end.model.DataManager
import me.imzack.app.end.model.bean.Type
import me.imzack.app.end.model.bean.TypeMarkPattern
import me.imzack.app.end.util.ResourceUtil
import me.imzack.app.end.view.adapter.TypeMarkPatternGridAdapter
import me.imzack.lib.basedialogfragment.BaseDialogFragment

class TypeMarkPatternPickerDialogFragment : BaseDialogFragment() {

    companion object {

        private val ARG_DEFAULT_PATTERN = "default_pattern"

        fun newInstance(defaultPattern: String?, typeMarkPatternPickedListener: (typeMarkPattern: TypeMarkPattern?) -> Boolean): TypeMarkPatternPickerDialogFragment {
            val fragment = TypeMarkPatternPickerDialogFragment()
            val args = Bundle()
            putBaseArguments(
                    args,
                    ResourceUtil.getString(R.string.title_dialog_fragment_type_mark_pattern_picker),
                    ResourceUtil.getString(R.string.button_remove),
                    object : OnButtonClickListener {
                        override fun onClick() = typeMarkPatternPickedListener(null)
                    },
                    ResourceUtil.getString(android.R.string.cancel),
                    object : OnButtonClickListener {
                        override fun onClick() = true
                    },
                    ResourceUtil.getString(R.string.button_select),
                    object : OnButtonClickListener {
                        // takeUnless 判断调用它的变量是否符合要求，如果符合，返回此变量，否则返回 null
                        override fun onClick() = typeMarkPatternPickedListener(fragment.typeMarkPattern.takeUnless { it.patternFn == null })
                    },
                    false
            )
            args.putString(ARG_DEFAULT_PATTERN, defaultPattern)
            fragment.arguments = args
            return fragment
        }
    }

    private val defaultPattern by lazy { arguments!!.getString(ARG_DEFAULT_PATTERN) }
    private val typeMarkPatternList = DataManager.typeMarkPatternList
    private var position = -1
    private val typeMarkPattern by lazy { if (position == -1) TypeMarkPattern() else typeMarkPatternList[position].copy() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // 为 position 设置正确的初始值
        if (Type.hasMarkPattern(defaultPattern)) {
            // indexOfFirst 找到第一个出现符合要求的元素下标，如果没找到，返回 -1（实际上不会出现）
            position = typeMarkPatternList.asSequence().indexOfFirst { it.patternFn == defaultPattern }
        }
    }

    override fun onCreateContentView(inflater: LayoutInflater, root: ViewGroup) =
            inflater.inflate(R.layout.dialog_fragment_type_mark_pattern_picker, root, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        grid_type_mark_pattern.adapter = TypeMarkPatternGridAdapter(typeMarkPatternList)

        if (position != -1) {
            grid_type_mark_pattern.setItemChecked(position, true)
        }

        grid_type_mark_pattern.onItemClickListener = AdapterView.OnItemClickListener { _, _, pos, _ ->
            position = pos
            val (patternFn, patternName) = typeMarkPatternList[position]
            typeMarkPattern.patternFn = patternFn
            typeMarkPattern.patternName = patternName
        }
    }
}
